packages/drivers/odsp-driver/src/compactSnapshotParser.ts

```diff
@@ -0,0 +1,156 @@
+    const dictionary = new Array<string>();
+    for (const name of node) {
+        assert(name instanceof BlobCore, "Mapping should be of type BlobCore");
+        dictionary.push(name.toString());
+    }
+    return dictionary;
+}
+
+/**
+ * Recreates blobs section of the tree.
+ * @param node - tree node to read blob section from
+ * @param dictionary - name map, used to decode path/IDs.
+ */
+function readBlobSection(node: NodeCore, dictionary: string[]) {
+    const blobs: Map<string, ArrayBuffer> = new Map();
+    for (const [idIndex, blob] of node.iteratePairs()) {
+        assert(typeof idIndex === "number", "Blob index should be a number");
+        assert(blob instanceof BlobCore, "Blob content should be of type blob");
+        const blobId = dictionary[idIndex];
+        assert(blobId !== undefined, "blob id should be present");
+        const unit8Array = blob.buffer;
+        blobs.set(blobId,
+            unit8Array.buffer.slice(unit8Array.byteOffset, unit8Array.byteOffset + unit8Array.byteLength));
+    }
+    return blobs;
+}
+
+/**
+ * Recreates ops section of the tree.
+ * @param node - tree node to read ops section from
+ */
+function readOpsSection(node: NodeCore) {
+    const ops: ISequencedDeltaOpMessage[] = [];
+    for (let i = 0; i < node.length; ++i) {
+        ops.push(JSON.parse(node.getString(i)));
+    }
+    return ops;
+}
+
+/**
+ * Recreates snapshot tree out of tree representation.
+ * @param treeNode - tree node to de-serialize from
+ * @param dictionary - name map, used to decode path/IDs.
+ */
+function readTreeSection(treeNode: NodeCore, dictionary: string[]) {
+    const tree: ISnapshotTree = {
+        blobs: {},
+        commits: {},
+        trees: {},
+    };
+    for (const [pathIndex, child] of treeNode.iteratePairs()) {
+        assert(typeof pathIndex == "number", "Tree index should be a number");
+        const path = dictionary[pathIndex];
+        assert(path !== undefined, "Path should not be undefined");
+
+        if (child instanceof NodeCore) {
+            tree.trees[path] = readTreeSection(child, dictionary);
+        } else {
+            assert(typeof child == "number", "Should be number to look in dictionary");
+            const id = dictionary[child];
+            assert(id !== undefined, "Id is out of range");
+            tree.blobs[path] = id;
+        }
+    }
+    return tree;
+}
+
+/**
+ * Converts snapshot from binary compact representation to tree/blobs/ops.
+ * @param buffer - Compact snapshot to be parsed into tree/blobs/ops.
+ * @returns - tree, blobs and ops from the snapshot.
+ */
+export function parseCompactSnapshotResponse(buffer: ReadBuffer) {
+    const builder = TreeBuilder.load(buffer);
+    let ops: ISequencedDeltaOpMessage[] | undefined;
+
+    assert(builder.length === 1, "1 root should be there");
+    const root = builder.getNode(0);
+    assert(root.length >= 4 && root.length <= 5, "4 or 5 sections should be there");
```

Comment:
I'd not limit it from top (5), to allow extensibility in the future. As long as bottom 5 are there, and signatures match, we should consume it.
Alternative design is possible and should be considered - every sub-tree has name (i.e. "trees", "blobs", "ops"), such that order does not matter, and then additions are simple in the future. I think we should strongly consider it.

Reply:
But it should be change in the server response that we get? Like currently we just rely on ordering in a  particular way and we have no way of knowing which section is which.

Reply:
Yep. We should not pass on opportunities here - if we agree it's good change, then we should make it on both client & server.

Reply:
I have talked to them about adding "string" properties in front of sections like blobs, trees, ops etc. After that we don't need to rely on length.